src/design_notes/fn_type_trait_impls.md

```diff
@@ -0,0 +1,258 @@
+  > because you (probably) would never actually use it really as a "default";
+  > e.g. Vec::resize'ing with it is super unlikely. It's also not really a
+  > Default but more just "the only value." Certainly the error message telling
+  > me that Default is not implemented for &fn() {foo} is likely to be pretty
+  > confusing since that does have a natural default too, like any pointer to
+  > ZST). That's in some sense just more broadly true though.
+
+- There were objections on the grounds that `Default` is not sufficient to
+  guarantee _uniqueness_ of the function value. Code could be written today that
+  exposes a public API with a `Default + Fn()` bound, expecting all types
+  meeting that bound to have a single unique value.
+
+  If we expanded the set of types which could implement `Default + Fn()` (such
+  as by stabilizing `Fn` trait implementations or by making more function
+  types implement `Default`) then the assumptions of such code would be
+  broken.
+
+  On the other hand, we really can't stop people from writing faulty code and
+  this does not seem like a footgun people are going to accidentally use, in
+  part because it's so obscure.
+
+### New lang-item
+
+This was a relatively minor consideration, but it is worth noting that this
+solution would require making `Default` a lang item.
+
+## Safe transmute
+
+This proposal was to materialize the closure using the machinery being
+added with the "safe transmute" RFC to transmute from the unit `()` type.
+
+The details of how this would work in practice were not discussed in detail,
+but there were some salient points:
+
+- This solves the "uniqueness" problem, in that ZSTs are by definition unique.
+- It does not help with the "privacy leakage" concerns.
+- It opens up a new can of worms relating to the fact that ZST closure types
+  may still have upvars.
+- Several people expressed something along the lines of:
+
+  > if we were going to have a trait that allows this, it might as well be
+  > Default, because telling people "no, you need the special default" doesn't
+  > really help anything.
+
+  Or, that if it's possible to do this one way with safe code, it should be
+  possible to do it in every way that makes sense.
+
+## `Singleton` or `ZST` trait
+
+New traits were proposed to avoid using `Default` to materialize the function
+values. The considerations here are mostly the same as for the "safe
+transmute" propsal. One note is that if we _were_ to add a `Singleton` trait,
+it would probably make sense for that trait to inherit from the `Default`
+trait anyway, and so a `Default` implementation now would be
+backwards-compatible.
+
+## `FnStatic` trait
+
+This would be a new addition to the set of `Fn` traits which would allow
+calling the function without any `self` argument at all. As the most
+restrictive (for the callee) and least restrictive (for the caller) it
+would sit at the bottom of the `Fn` trait hierarchy and inherit from `Fn`.
```

Comment:
This would also make the zst closure -> fn ptr coercion in rustc less hacky. It currently depends on the fact that the self argument of the `FnOnce::call_once` method is skipped in the ABI calculation because it is a ZST. This means that the zst closure -> fn ptr coercion can transmute `FnOnce::call_once` to a function pointer without the self argument. With `FnStatic`, the coercion could just return `FnStatic::call_static`.

Reply:
I think it's worth nothing that, or at least the "precedent" of this coercion (not sure if we had it already)

Reply:
I've added a note on this.